src/ScheduleFunctions.cpp

```diff
@@ -414,288 +408,263 @@ Stmt build_produce(const map<string, Function> &env, Function f, const Target &t
+            // The cropped buffer_t
+            cleanup_args.push_back(p.first);
+            // Its parent
+            cleanup_args.push_back(p.second);
         }
 
-        for (size_t i = 0; i < lets.size(); i++) {
-            check = LetStmt::make(lets[i].first, lets[i].second, check);
+        if (cropped_buffers.size() > 1) {
+            // NULL-terminate it
+            cleanup_args.push_back(make_zero(type_of<struct halide_buffer_t *>()));
         }
 
-        return build_loop_nest(check, f.name() + ".s0.", -1, f.args(), {}, f.schedule(), f.definition().schedule(), false);
-    } else {
+        Expr cleanup_struct = Call::make(Handle(),
+                                         Call::make_struct,
+                                         cleanup_args,
+                                         Call::Intrinsic);
 
-        string prefix = f.name() + ".s0.";
-        vector<string> dims = f.args();
-        return build_provide_loop_nest(env, f.name(), prefix, -1, dims, f.schedule(), f.definition(), false);
+        // We have to anticipate failures in the extern stage, so
+        // call this by registering a destructor before the extern
+        // call and triggering it afterwards using an Allocate node.
+        string destructor_name = unique_name('d');
+        const char *fn = (cropped_buffers.size() == 1 ?
+                          "_halide_buffer_retire_crop_after_extern_stage" :
+                          "_halide_buffer_retire_crops_after_extern_stage");
+        check = Allocate::make(destructor_name, Handle(), MemoryType::Heap, {},
+                               const_true(), check, cleanup_struct, fn);
     }
-}
-
-// Build the loop nests that update a function (assuming it's a reduction).
-vector<Stmt> build_update(const map<string, Function> &env, Function f) {
-
-    vector<Stmt> updates;
 
-    for (size_t i = 0; i < f.updates().size(); i++) {
-        const Definition &def = f.update(i);
-
-        string prefix = f.name() + ".s" + std::to_string(i+1) + ".";
-
-        vector<string> dims = f.args();
-        Stmt loop = build_provide_loop_nest(env, f.name(), prefix, -1, dims, f.schedule(), def, true);
-        updates.push_back(loop);
+    if (annotate.defined()) {
+        check = Block::make(annotate, check);
     }
 
-    return updates;
-}
-
-pair<Stmt, Stmt> build_production(const map<string, Function> &env, Function func, const Target &target) {
-    Stmt produce = build_produce(env, func, target);
-    vector<Stmt> updates = build_update(env, func);
+    for (const auto &let : lets) {
+        check = LetStmt::make(let.first, let.second, check);
+    }
 
-    // Combine the update steps
-    Stmt merged_updates = Block::make(updates);
-    return { produce, merged_updates };
+    Definition f_def_no_pred = f.definition().get_copy();
+    f_def_no_pred.predicate() = const_true();
+    return build_loop_nest(check, f.name() + ".s0.", -1, f, f_def_no_pred, false);
```

Comment:
This line used to be the following in #3352 
```
return build_loop_nest(check, f.name() + ".s0.", -1, f.args(), {}, f.schedule(), f.definition().schedule(), false);
```
I'm not sure why the split predicates list is dropped going into this call, but I replicated that behavior. Could someone shed some light on this? Maybe we could put in a comment.